### PR TITLE
Feat: authenticated calls in Candid UI

### PR DESCRIPTION
**Pull request by Web3NL**
See [#474](https://github.com/dfinity/candid/issues/474) and [demo link](https://m4ul7-aqaaa-aaaal-qcewq-cai.raw.icp0.io/?id=mvxad-wyaaa-aaaal-qcexa-cai)

**Overview**
Allow authenticated calls via Internet Identity (and possibly pem files)

**Requirements**
New deps: `auth-client` and `identity` packages from `agent-js`
Modifications and additions to `tools/ui`

**Considered Solutions**
Added II to Candid UI canister

**Recommended Solution**
Also recommend uploading pem files (with warning to developer about security)

**Considerations**
Should be easy to maintain, modifications of existing files are kept to a minimum
New deps version bumps

**TODO**
Remove `dfx.json`, `canister_ids.json` and `target_test_canister` from branch (left them for demo purpose)
Browser compatibility check
Testing preferably with Docker and Selenium (for local II alternative origins testing over https)
UI enhancements
Gather feedback

### CODE REVIEW

**Comment by rvanasa** [comment]
Looks good so far! Thanks for the contribution!

File: tools/ui/src/candid.ts [hunk 1]
```diff
@@ -77,6 +80,14 @@ export async function fetchActor(canisterId: Principal): Promise<ActorSubclass>
   }
   const dataUri = 'data:text/javascript;charset=utf-8,' + encodeURIComponent(js);
   const candid: any = await eval('import("' + dataUri + '")');
+
+  authClient = authClient ?? (await AuthClient.create())
```
Comment (rvanasa) [inline tools/ui/src/candid.ts:84]: Perhaps we could deactivate the idle timeout (similar to the NNS dapp):

```suggestion
  authClient = authClient ?? (await AuthClient.create({
    idleOptions: {
      disableIdle: true,
      disableDefaultIdleCallback: true,
    },
  }))
```

**Review by v2i0s2h2 — COMMENTED** [review]
sir

**Review by maheshVishwakarma1998 — COMMENTED** [review]
ser

File: tools/ui/dfx.json [hunk 2]
```diff
@@ -14,6 +14,11 @@
           "name": "candid:service"
         }
       ]
+    },
+    "target": {
```
Comment (chenyan-dfinity) [inline tools/ui/dfx.json:18]: We don't need this canister.

**Review by chenyan-dfinity — APPROVED** [review]
Thanks for your contribution! 

We don't need the `target_test_canister` here. I can merge once it's removed.

**Comment by Web3NL** [comment]
Thank you!

File: tools/ui/target_test_canister/AlternativeOrigin.mo [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import Text "mo:base/Text";
```
Comment (chenyan-dfinity) [inline tools/ui/target_test_canister/AlternativeOrigin.mo:1]: Also remove this directory.

**Comment by Web3NL** [comment]
Here we go!

**Comment by Web3NL** [comment]
Many thnx!

Will this make it into dfx 0.15.2? I would be so proud 😄

**Comment by chenyan-dfinity** [comment]
@Web3NL Actually, there is one more UI fix we need to do. We use Candid UI in the playground with small width. The II button looks big there. Is there an easy way to make it smaller with small window width? (cc @rvanasa if you have a quick fix)

Screenshot 2023-10-17 at 12 13 13 PM

**Comment by Web3NL** [comment]
Will have a look tomorrow i hope

**Comment by ggreif** [comment]
@Web3NL Does the merged version work with `ic0.app`-domain anchors too?

**Comment by chenyan-dfinity** [comment]
Yes, I think it supports both `raw` and `icp0.io`

**Comment by Web3NL** [comment]
> @Web3NL Does the merged version work with `ic0.app`-domain anchors too?

yes, i added a form to choose between ic0.app / icp0.io. Same for raw / non-raw

**Comment by Web3NL** [comment]
> @Web3NL Does the merged version work with `ic0.app`-domain anchors too?

Excuse me, I corrected my last response.

Does this answer your question, or were you refering to the default identity provider, which now is always `identity.internetcomputer.org` and never `identity.ic0.app`

The `derivationOrigin` (canister for which we want to obtain a Principal) may be `icp0.io` or `ic0.app`

**Comment by rvanasa** [comment]
> Actually, there is one more UI fix we need to do. We use Candid UI in the playground with small width. The II button looks big there. Is there an easy way to make it smaller with small window width?

From a responsive UX standpoint, the following three improvements would help a lot on smaller screen sizes:

- Using `@media` CSS queries to change the login form / button layout for different screen widths
- Moving checkboxes inside of the `<label>` tag (which makes it possible to toggle the checkbox by clicking the label)
- Shortening "Login with Internet Identity" to just "Login" (maybe depending on screen width)

@Web3NL, here is a partial solution in case you want to borrow any of this code: https://github.com/dfinity/candid/pull/480

**Comment by Web3NL** [comment]
I will have a look, thnx for tagging.

And I think this feature deserves some documention on the portal on how a canister should serve the `ii-alternative-origins` file.

I will attend to that 🙂

**Comment by ggreif** [comment]
> yes, i added a form to choose between ic0.app / icp0.io. Same for raw / non-raw
> Note, the obtained principal is independent of derivation origin domain, but does change if you include / exclude raw

I am asking because my `ic0.app`-based anchor didn't work in the example canister.

**Comment by Web3NL** [comment]
Ah, I see.

Is this because `identityProvider` is set to `identity.internetcomputer.org`?

If so, then there should also be an option to choose `identity.ic0.app` as the provider

**Comment by chenyan-dfinity** [comment]
Deployed to the main net!

Another possible improvement is to only show the `Login` button when `alternativeOrigins` is presented in the destination canister response.

**Pull request merged**